### ISSUE TITLE
Cannot generate JSON schema anymore with `Literal[MyEnum.MEMBER]`

### ISSUE THREAD

**Issue by nymous** [samuelcolvin/pydantic#2534]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

I found what seems to be a regression in Pydantic 1.8+.

When a model has property typed as a `Literal[MyEnum.MEMBER]`, the `schema()` and `schema_json()` methods fail, until I add another property typed as `MyEnum`.

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: True
                 install path: /tmp/tmp.iid8D0iv22/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.2 (default, Feb 20 2021, 18:40:11)  [GCC 10.2.0]
                     platform: Linux-5.11.6-arch1-1-x86_64-with-glibc2.33
     optional deps. installed: ['typing-extensions']
```

```py
from typing import Literal
from enum import Enum

from pydantic import BaseModel

class MyEnum(str, Enum):
    FOO = "foo"
    BAR = "bar"


class ModelWithLiteralEnumMember(BaseModel):
    kind: Literal[MyEnum.FOO]

ModelWithLiteralEnumMember.schema_json()  # This works until 1.7.3 but crashes on 1.8+


class ModelWithLiteralEnumMemberAndFullEnum(BaseModel):
    kind: Literal[MyEnum.FOO]
    possible_kinds: MyEnum

ModelWithLiteralEnumMemberAndFullEnum.schema_json()  # ... but now with the new property it does work
```

Traceback looks like this:
```
Traceback (most recent call last):
  File "/tmp/tmp.iid8D0iv22/main.py", line 33, in <module>
    print(ModelWithLiteralEnumMember.schema_json())  # This works until 1.7.3 but crashes on 1.8+
  File "pydantic/main.py", line 715, in pydantic.main.BaseModel.schema_json
  File "pydantic/main.py", line 704, in pydantic.main.BaseModel.schema
  File "pydantic/schema.py", line 167, in pydantic.schema.model_schema
  File "pydantic/schema.py", line 548, in pydantic.schema.model_process_schema
  File "pydantic/schema.py", line 589, in pydantic.schema.model_type_schema
  File "pydantic/schema.py", line 241, in pydantic.schema.field_schema
  File "pydantic/schema.py", line 495, in pydantic.schema.field_type_schema
  File "pydantic/schema.py", line 810, in pydantic.schema.field_singleton_schema
KeyError: <enum 'MyEnum'>
```

I have made a more complete gist comparing the results in 1.7.3, 1.8 and 1.8.1, with multiple tries to workaround the issue: https://gist.github.com/nymous/686d38ba1aa150fefe6da6d78b1cb441

cc @Senpos, what we discussed on the FastAPI Discord

**Comment by PrettyWood** [comment]
Hi @nymous and thanks for reporting! 👍 
You're right there is a tiny error that I just fixed!
Can you please check that this indeed solves both your issues? 🙏

**Comment by nymous** [comment]
Wow that is a fast fix!
I just tested with a sample FastAPI app (that was where I found the bug initially), works flawlessly!
Thank you so much!
```py
from enum import Enum
from typing import Literal, Union

from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class ServerType(str, Enum):
    FOO = "foo"
    BAR = "bar"

class ServerBase(BaseModel):
    type: ServerType

class FooServer(ServerBase):
    type: Literal[ServerType.FOO]
    foo_field: str

class BarServer(ServerBase):
    type: Literal[ServerType.BAR]
    bar_field1: int
    bar_field2: float


@app.post("/server")
async def index(server_info: Union[FooServer, BarServer]):
    return server_info
```

**Comment by kataev** [comment]
well, any news?

**Comment by PrettyWood** [comment]
I created a milestone for a v1.8.2. @samuelcolvin Do you think we could ship one?

**Comment by samuelcolvin** [comment]
Thanks so much for reporting this, the PR is merged and should be included in 1.9 which I'm working to get out now.

I'm sorry about v1.8.2, I never really got around to reviewing all the required PRs and now I think it's too late, hopefully 1.9 fixes the problems - it should be smaller than 1.8 so is less likely to include breaking changes itself.
